A68-random-subsetting.md

```diff
@@ -0,0 +1,117 @@
+A68: Random subsetting with rendezvous hashing LB policy.
+----
+* Author(s): @s-matyukevich
+* Approver: 
+* Status: Draft
+* Implemented in: Go, Java
+* Last updated: [Date]
+* Discussion at: 
+
+## Abstract
+
+Add support for the `random_subsetting` load balancing policy.
+
+## Background
+
+Currently, gRPC is lacking a way to select a subset of endpoints available from the resolver and load-balance requests between them. Out of the box, users have the choice between two extremes: `pick_first` which sends all requests to one random backend, and `round_robin` which sends requests to all available backends. `pick_first` has poor connection balancing when the number of client is not much higher than the number of servers. The problem is exacerbated during rollouts because `pick_first` does not change endpoint on resolver updates if the current subchannel remains `READY`. `round_robin` results in every server having as many connections open as there are clients, which is unnecessarily costly when there are many clients, and makes local load balancing decisions (such as outlier detection) less precise.
+
+### Related Proposals: 
+* [gRFC A27: A52: gRPC xDS Custom Load Balancer Configuration](https://github.com/grpc/proposal/blob/master/A52-xds-custom-lb-policies.md)
+
+## Proposal
+
+Introduce a new LB policy, `random_subsetting`. This policy selects a subset of addresses and passes them to the child LB policy. It maintains 2 important properties:
+* The policy tries to distribute connections among servers as equally as possible. The higher `(N_clients*subset_size)/N_servers` ratio is, the closer the resulting server connection distribution is to uniform.
+* The policy minimizes the amount of connection churn generated during server scale-ups by using [rendezvous hashing](https://en.wikipedia.org/wiki/Rendezvous_hashing)
+
+### Subsetting algorithm
+
+* The policy receives a single configuration parameter: `subset_size`, which must be configured by the user.
+* When the lb policy is initialized it also creates a random 32-byte long `salt` string. 
+* After every resolver update the policy picks a new subset. It does this by implementing `rendezvous hashing` algorithm:
+  * Concatenate string representation of every address in the list with `salt`.
```

Comment:
I misunderstood this sentence at first, I thought you meant to concatenate all the addresses together and then add the salt. I think making it more explicit would be easier to understand for the reader, eg. `Concatenate `salt` to each address in the list`.

Also might be good to specify what string representation to use for an IP address as multiple exist.

Reply:
updated this bullet point. 

> Also might be good to specify what string representation to use for an IP address as multiple exist.

This might be language dependent, but we don't care which one we should use. It just must be stable for a given client.

Reply:
Got it, so multiple clients using a different string representation is not an issue?

Also, would this be a problem for a client:

```python
>>> import ipaddress
>>> ipaddress.ip_address("fe80::1.2.3.4")
IPv6Address('fe80::102:304')
```

The string representation of this IPv6 address can be in multiple forms. They are both valid and both represent the same 64 bit integer address but a common bug is for application code to think they are distinct addresses. I think it's fine for your use-case but wanted to validate it.